internal/client/client.go

```diff
@@ -20,40 +20,60 @@ import (
 var UserAgent = fmt.Sprintf("Go-Auth0-SDK/%s", auth0.Version)
 
 func WrapRateLimit(c *http.Client) *http.Client {
-	return &http.Client{
-		Transport: rehttp.NewTransport(
-			c.Transport,
-			func(attempt rehttp.Attempt) bool {
-				if attempt.Response == nil {
-					return false
-				}
-				return attempt.Response.StatusCode == http.StatusTooManyRequests
-			},
-			func(attempt rehttp.Attempt) time.Duration {
-				resetAt := attempt.Response.Header.Get("X-RateLimit-Reset")
-				resetAtUnix, err := strconv.ParseInt(resetAt, 10, 64)
-				if err != nil {
-					resetAtUnix = time.Now().Add(5 * time.Second).Unix()
-				}
-				return time.Duration(resetAtUnix-time.Now().Unix()) * time.Second
-			},
-		),
-	}
+	c.Transport = rehttp.NewTransport(
+		c.Transport,
+		func(attempt rehttp.Attempt) bool {
+			if attempt.Response == nil {
+				return false
+			}
+			return attempt.Response.StatusCode == http.StatusTooManyRequests
+		},
+		func(attempt rehttp.Attempt) time.Duration {
+			resetAt := attempt.Response.Header.Get("X-RateLimit-Reset")
+			resetAtUnix, err := strconv.ParseInt(resetAt, 10, 64)
+			if err != nil {
+				resetAtUnix = time.Now().Add(5 * time.Second).Unix()
+			}
+			return time.Duration(resetAtUnix-time.Now().Unix()) * time.Second
+		},
+	)
+	return c
 }
 
-func WrapUserAgent(c *http.Client, userAgent string) *http.Client {
-	return &http.Client{
-		Transport: RoundTripFunc(func(req *http.Request) (*http.Response, error) {
-			req.Header.Set("User-Agent", userAgent)
-			return c.Transport.RoundTrip(req)
-		}),
+// Transport is used to wrap http.Client.Transport when other wrap functions make changes to http.Client.
+type Transport struct {
+	base    http.RoundTripper
+	headers map[string]string
+	debug   bool
+}
+
+// RoundTrip ensures Transport satisfies http.RoundTripper adding the ability to set headers and enable debugging.
+func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	for key, value := range t.headers {
+		req.Header.Set(key, value)
 	}
+	if t.debug {
+		dumpRequest(req)
+		res, err := t.base.RoundTrip(req)
+		if err != nil {
+			return res, err
+		}
+		dumpResponse(res)
+		return res, nil
+	}
+	return t.base.RoundTrip(req)
 }
 
-type RoundTripFunc func(*http.Request) (*http.Response, error)
+func newTransport(base http.RoundTripper, headers map[string]string, debug bool) *Transport {
+	if base == nil {
+		base = http.DefaultTransport
+	}
+	return &Transport{base, headers, debug}
+}
 
-func (rf RoundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
-	return rf(req)
+func WrapUserAgent(c *http.Client, userAgent string) *http.Client {
+	c.Transport = newTransport(c.Transport, map[string]string{"User-Agent": userAgent}, false)
+	return c
 }
```

Comment:
To support the functional arguments passed into `Client(...)` we'll have to change the signature (and name) to allow this:

```go
func WithRateLimit() func(*http.Client) { ... }
```

```suggestion
// WithUserAgent wraps c.Transport in a way that adds a User-Agent header to
// each request.
func WithUserAgent(userAgent string) func(*http.Client) {
	return func(c *http.Client) {
		baseTransport := c.Transport
		c.Transport = RoundTripFunc(func(req *http.Request) (*http.Response, error) {
			req.Header.Set("User-Agent", userAgent)
			return baseTransport.RoundTrip(req)
		})
	}
}
```

The same should happen for `WrapDebug` and `WrapRateLimit`.